app/database2.js

```diff
@@ -0,0 +1,306 @@
+	var sql,db,db_online;
+	if (typeof sqlite !== 'object') {
+		document.body.style.backgroundColor = 'red';
+		alert("Failed to require sqlite through AMD.");
+	} else {
+		sql = sqlite;
+	}
+
+	// Initiate DB and check if there is an existing user DB
+	var read_database;
+	if (database_exists(config.constant("DATABASE_USER"))) {
+		read_database = fs.readFileSync(config.constant("DATABASE_USER"));
+	} else {
+		read_database = fs.readFileSync(config.constant("DATABASE_SLIMSTAMPEN"));
+	}
+	var db = new sql.Database(read_database);
+
+	function database_exists(path) {
+		try {
+			fs.accessSync(path, fs.F_OK);
+			return true;
+		} catch (e) {
+			return false;
+		}
+	}
+
+	function onError(error) {
+		if (error) {
+			if (error.message !== undefined) {
+				console.log(error.message);
+			} else {
+				console.log(error);
+			}
+		}
+	}
+	function isUnique(unique_name1, unique_name2, queryResult, row) {
+	    for (i = 0; i<queryResult.length;i++) {
+	      if (queryResult[i][unique_name1]==row[unique_name1] && queryResult[i][unique_name2]==row[unique_name2]) {
+					return false;
+	      }
+	    }
+	    return true;
+	}
+  function isUniqueOld(unique_name, queryResult, row) {
+    for (i = 0; i<queryResult.length;i++) {
+      if (queryResult[i][unique_name]==row[unique_name]) {
+        return false;
+      }
+    }
+    return true;
+  }
+	function synchronizeUser(userId, callback) {
+		var local_user = database.getQuery('getUser',[userId]);
+		database.getOnlineQuery("getUser", [userId], function(online_user) {
+			online_user = online_user[0];
+			if (local_user.length === 0) {
+				online_user = $.map(online_user, function(val, key) { return (key=="user_createdate" || key=="user_lastedited" || key=="user_bday") ? date.formatDatetime(val) : val; });
+				database.executeQuery('addUser', online_user, true, false);
+			} else {
+				local_user = local_user[0];
+				var localTime = Date.parse(local_user.user_lastedited);
+				var onlineTime = Date.parse(online_user.user_lastedited);
+				var recent = localTime - onlineTime;
+				if (recent > 0) {
+					local_user = $.map(local_user, function(val, key) { return (key=="user_createdate" || key=="user_lastedited" || key=="user_bday") ? date.formatDatetime(val) : val; });
+					database.executeQuery('replaceUser', local_user, false, true, function() {
+						callback();
+					});
+				} else if (recent < 0) {
+					online_user = $.map(online_user, function(val, key) { return (key=="user_createdate" || key=="user_lastedited" || key=="user_bday") ? date.formatDatetime(val) : val; });
+					database.executeQuery('replaceUser', online_user, true, false);
+				}
+			}
+			callback();
+		});
+	}
+
+	function synchronizeDatasets(userId, callback) {
+			synchronizing = true;
+			var localdatasets = database.getQuery('getUserDatasets',[userId]);
+			var remotetolocaldatasets = [];
+			database.getOnlineQuery('getUserDatasets',[userId], function(remotedatasets) {
```

Comment:
Too many things on too many abstraction levels, include some functions, let the comments guide you.